File: tools/hls-edge/test/ll-state.test.mjs
import { strict as assert } from "node:assert";
import test from "node:test";

import {
  LL_AUDIO_RUNG,
  LL_VIDEO_RUNG,
  parseLlState,
  playlistOriginKindForRung,
  trackForRung,
} from "../src/ll-state.js";

/** A minimal, valid state.json body: two complete segments, one partial, no audio. */
function fixtureState(overrides = {}) {
  return {
    sessionId: "5a1b2c3d-1234-5678-9abc-1234567890ab",
    channelId: "chan_abc123",
    partTargetMs: 500,
    segmentTargetMs: 4000,
    targetDurationSecs: 4.5,
    mediaSequence: 41,
    video: {
      initUri: "init.mp4",
      segments: [
        {
          msn: 41,
          complete: true,
          durationSecs: 4.016,
          programDateTime: "2026-09-14T18:03:21.114Z",
          uri: "seg-41.m4s",
          parts: [
            { index: 0, durationSecs: 0.501, independent: true, uri: "part-41.0.m4s" },
            { index: 1, durationSecs: 0.498, independent: false, uri: "part-41.1.m4s" },
          ],
        },
        {
          msn: 42,
          complete: true,
          durationSecs: 4.0,
          programDateTime: "2026-09-14T18:03:25.130Z",
          uri: "seg-42.m4s",
          parts: [{ index: 0, durationSecs: 0.5, independent: true, uri: "part-42.0.m4s" }],
        },
        {
          msn: 43,
          complete: false,
          programDateTime: "2026-09-14T18:03:29.130Z",
          parts: [{ index: 0, durationSecs: 0.502, independent: true, uri: "part-43.0.m4s" }],
        },
      ],
      preloadHint: { msn: 43, part: 1, uri: "part-43.1.m4s" },
    },
    audio: null,
    ...overrides,
  };
}

test("parses a well-formed fixture", () => {
  const state = parseLlState(fixtureState());
  assert.ok(state);
  assert.equal(state.sessionId, "5a1b2c3d-1234-5678-9abc-1234567890ab");
  assert.equal(state.video.segments.length, 3);
  assert.equal(state.video.segments[2].complete, false);
  assert.equal(state.video.preloadHint.uri, "part-43.1.m4s");
  assert.equal(state.audio, null);
});

test("rejects a non-object", () => {
  assert.equal(parseLlState(null), null);
  assert.equal(parseLlState("not json"), null);
  assert.equal(parseLlState(42), null);
});

test("rejects an empty segments array", () => {
  const raw = fixtureState();
  raw.video.segments = [];
  assert.equal(parseLlState(raw), null);
});

test("rejects an incomplete segment anywhere but last", () => {
  const raw = fixtureState();
  raw.video.segments[0].complete = false;
  delete raw.video.segments[0].uri;
  assert.equal(parseLlState(raw), null);
});

test("rejects a complete segment missing its duration or uri", () => {
  const missingDuration = fixtureState();
  delete missingDuration.video.segments[0].durationSecs;
  assert.equal(parseLlState(missingDuration), null);

  const missingUri = fixtureState();
  delete missingUri.video.segments[0].uri;
  assert.equal(parseLlState(missingUri), null);
});

test("rejects an incomplete segment that claims a uri", () => {
  const raw = fixtureState();
  raw.video.segments[2].uri = "seg-43.m4s";
  assert.equal(parseLlState(raw), null);
});

test("rejects a part missing independent/uri/durationSecs", () => {
  const raw = fixtureState();
  delete raw.video.segments[0].parts[0].independent;
  assert.equal(parseLlState(raw), null);
});

test("rejects a bad programDateTime", () => {
  const raw = fixtureState();
  raw.video.segments[0].programDateTime = "not a date";
  assert.equal(parseLlState(raw), null);
});

test("preloadHint may be null, and a malformed one is rejected", () => {
  const nullHint = fixtureState();
  nullHint.video.preloadHint = null;
  assert.ok(parseLlState(nullHint));

  const badHint = fixtureState();
  badHint.video.preloadHint = { msn: 43 }; // missing part/uri
  assert.equal(parseLlState(badHint), null);
});

test("accepts a state with an audio track and trackForRung resolves both rungs", () => {
  const raw = fixtureState({
    audio: {
      initUri: "audio-init.mp4",
      segments: [
        {
          msn: 41,
          complete: true,
          durationSecs: 4.0,
          programDateTime: "2026-09-14T18:03:21.114Z",
          uri: "audio-seg-41.m4s",
          parts: [{ index: 0, durationSecs: 0.5, independent: true, uri: "audio-part-41.0.m4s" }],
        },
      ],
      preloadHint: null,
    },
  });
  const state = parseLlState(raw);
  assert.ok(state);
  assert.equal(trackForRung(state, LL_VIDEO_RUNG), state.video);
  assert.equal(trackForRung(state, LL_AUDIO_RUNG), state.audio);
  assert.equal(trackForRung(state, "720p30"), null);
});

test("a malformed audio track fails the whole document, not just the audio half", () => {
  const raw = fixtureState({ audio: { initUri: "audio-init.mp4", segments: [] } });
  assert.equal(parseLlState(raw), null);
});

test("rejects non-positive partTargetMs/segmentTargetMs/targetDurationSecs", () => {
  for (const field of ["partTargetMs", "segmentTargetMs", "targetDurationSecs"]) {
    const raw = fixtureState({ [field]: 0 });
    assert.equal(parseLlState(raw), null, `field ${field}`);
  }
});

test("rejects a sessionId that is not shaped like a UUID -- unescaped into playlist text and used as a request path segment", () => {
  for (const badId of ["not-a-uuid", "5a1b2c3d\n#EXT-X-DISCONTINUITY", "../../evil", ""]) {
    const raw = fixtureState({ sessionId: badId });
    assert.equal(parseLlState(raw), null, `sessionId ${JSON.stringify(badId)}`);
  }
});

test("rejects an otherwise-valid sessionId/uri with a trailing newline appended -- \"value\\n\" must not slip past the $ anchor", () => {
  const validSessionId = "5a1b2c3d-1234-5678-9abc-1234567890ab";
  for (const suffix of ["\n", "\r\n", "\r", "\n\n"]) {
    const raw = fixtureState({ sessionId: validSessionId + suffix });
    assert.equal(parseLlState(raw), null, `sessionId with suffix ${JSON.stringify(suffix)}`);
  }

  const validUri = "seg-41.m4s";
  for (const suffix of ["\n", "\r\n", "\r", "\n\n"]) {
    const raw = fixtureState();
    raw.video.segments[0].uri = validUri + suffix;
    assert.equal(parseLlState(raw), null, `uri with suffix ${JSON.stringify(suffix)}`);
  }
});

test("rejects a uri containing path separators, .., or control characters (playlist injection / origin-fetch escape)", () => {
  for (const badUri of ["../../secret", "a/b.m4s", "seg 41.m4s", "seg\n41.m4s", "http://evil.example/x", ""]) {
    const badInitUri = fixtureState();
    badInitUri.video.initUri = badUri;
    assert.equal(parseLlState(badInitUri), null, `initUri ${JSON.stringify(badUri)}`);

    const badSegmentUri = fixtureState();
    badSegmentUri.video.segments[0].uri = badUri;
    assert.equal(parseLlState(badSegmentUri), null, `segment uri ${JSON.stringify(badUri)}`);

    const badPartUri = fixtureState();
    badPartUri.video.segments[0].parts[0].uri = badUri;
    assert.equal(parseLlState(badPartUri), null, `part uri ${JSON.stringify(badUri)}`);
  }
});

test("rejects non-contiguous or duplicate segment MSNs", () => {
  const gap = fixtureState();
  gap.video.segments[1].msn = 45; // was 42, leaves a gap after 41
  assert.equal(parseLlState(gap), null);

  const duplicate = fixtureState();
  duplicate.video.segments[1].msn = 41; // duplicates segment 0's msn
  assert.equal(parseLlState(duplicate), null);

  const outOfOrder = fixtureState();
  outOfOrder.video.segments.reverse();
  assert.equal(parseLlState(outOfOrder), null);
});

test("rejects mediaSequence that does not match the video track's oldest segment MSN", () => {
  const raw = fixtureState({ mediaSequence: 40 }); // fixture's first video segment is msn 41
  assert.equal(parseLlState(raw), null);
});

test("rejects a complete segment whose duration exceeds the rendered target duration by more than the jitter tolerance", () => {
  const raw = fixtureState({ targetDurationSecs: 1 }); // segment 0 claims durationSecs 4.016, way past 1 + 0.5
  assert.equal(parseLlState(raw), null);
});

test("accepts a segment slightly past a whole-second targetDurationSecs -- ordinary encoder jitter, not a malformed snapshot", () => {
  // The exact shape RFC 8216bis players already tolerate: a 2.0s target
  // renders EXT-X-TARGETDURATION:2, and a real segment landing at 2.04s
  // (a few ms of encoder/keyframe jitter) still renders a legal EXTINF.
  // Comparing the segment's raw duration against the RAW targetDurationSecs
  // (rather than the rendered, ceiling'd value, with a small allowance for
  // this exact kind of jitter) rejected this legitimate snapshot outright.
  const raw = fixtureState({ targetDurationSecs: 2 });
  raw.video.segments[0].durationSecs = 2.04;
  // segments[1] and the live segment/parts must stay internally consistent
  // (contiguous MSNs, preloadHint placement) -- only segment 0's duration
  // is under test here, so shrink the others to fit comfortably under the
  // same target too.
  raw.video.segments[1].durationSecs = 2.0;
  assert.ok(parseLlState(raw));
});

test("still rejects a segment many multiples of the target -- the tolerance does not swallow a genuinely malformed duration", () => {
  const raw = fixtureState({ targetDurationSecs: 2 });
  raw.video.segments[0].durationSecs = 20; // 10x the target, not encoder jitter
  assert.equal(parseLlState(raw), null);
});

test("rejects a live segment with zero parts and no preload hint -- nothing published, nothing scheduled", () => {
  const raw = fixtureState();
  raw.video.segments[2].parts = [];
  raw.video.preloadHint = null;
  assert.equal(parseLlState(raw), null);

  // The same zero-part live segment IS accepted once a preload hint names
  // the very next part.
  const withHint = fixtureState();
  withHint.video.segments[2].parts = [];
  withHint.video.preloadHint = { msn: 43, part: 0, uri: "part-43.0.m4s" };
  assert.ok(parseLlState(withHint));
});

test("rejects a preload hint that re-announces an already-published part", () => {
  const raw = fixtureState();
  // segments[2] (msn 43) already has part index 0 published -- a hint for
  // {msn: 43, part: 0} re-announces it instead of naming the next one.
  raw.video.preloadHint = { msn: 43, part: 0, uri: "part-43.0.m4s" };
  assert.equal(parseLlState(raw), null);
});

test("rejects a preload hint pointing at the wrong segment once the live one is sealed", () => {
  const raw = fixtureState();
  raw.video.segments[2].complete = true;
  raw.video.segments[2].durationSecs = 4.0;
  raw.video.segments[2].uri = "seg-43.m4s";
  // Now that msn 43 is sealed, a valid hint must be {msn: 44, part: 0} --
  // this one still points at the old live edge.
  raw.video.preloadHint = { msn: 43, part: 1, uri: "part-43.1.m4s" };
  assert.equal(parseLlState(raw), null);

  const correct = fixtureState();
  correct.video.segments[2].complete = true;
  correct.video.segments[2].durationSecs = 4.0;
  correct.video.segments[2].uri = "seg-43.m4s";
  correct.video.preloadHint = { msn: 44, part: 0, uri: "part-44.0.m4s" };
  assert.ok(parseLlState(correct));
});

test("playlistOriginKindForRung: only the two LL rung names route to the LL origin", () => {
  assert.equal(playlistOriginKindForRung(LL_VIDEO_RUNG), "ll");
  assert.equal(playlistOriginKindForRung(LL_AUDIO_RUNG), "ll");
  // Every conventional rendition name in this codebase, plus garbage — all
  // of them must stay on "api", the SAME origin and code path they always
  // used. This is the property behind "conventional sessions are untouched
  // byte-for-byte" (docs/plans/LL_HLS.md task L2.2).
  for (const rung of ["720p30", "1080p60", "audio", "ll2", "ll-audio2", "", "LL", "LL-AUDIO"]) {
    assert.equal(playlistOriginKindForRung(rung), "api", `expected "api" for rung ${JSON.stringify(rung)}`);
  }
});

File: tools/pqp-remux/internal/llstate/llstate_test.go
package llstate

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"

	"github.com/rafaelcg/pqp/tools/pqp-remux/internal/pipeline"
	"github.com/rafaelcg/pqp/tools/pqp-remux/internal/ring"
)

var updateGolden = flag.Bool("update-golden", false,
	"rewrite testdata/state-golden.json (and therefore what tools/hls-edge's cross-check test reads) from this package's own renderer")

const (
	videoTimescale = 90000
	audioTimescale = 48000
	partTicks      = videoTimescale / 2 // 500ms
)

var epoch = time.Date(2026, 9, 15, 8, 1, 0, 0, time.UTC)

// videoSnapshot builds a synthetic ring snapshot: `sealed` segments of
// `partsPerSegment` parts each starting at segment index `first`, then one
// open segment with `openParts` parts. Part sequence numbers run 1..N
// across the whole track, exactly as pipeline.Fragmenter numbers them.
func videoSnapshot(first, sealed, partsPerSegment, openParts int, timescale uint32, ticks uint32) ring.Snapshot {
	snap := ring.Snapshot{Timescale: timescale, HasInit: true}
	seq := uint32(0)
	longest := 0.0
	partDur := time.Duration(float64(ticks) / float64(timescale) * float64(time.Second))
	elapsed := time.Duration(0)
	add := func(index, parts int, isSealed bool) {
		seg := ring.SegmentSnapshot{
			Index:    index,
			Sealed:   isSealed,
			OpenedAt: epoch.Add(elapsed),
		}
		elapsed += time.Duration(parts) * partDur
		for i := 0; i < parts; i++ {
			seq++
			seg.Parts = append(seg.Parts, ring.PartSnapshot{
				Seq:           seq,
				DurationTicks: ticks,
				Independent:   i == 0,
			})
		}
		if isSealed {
			if d := float64(parts) * float64(ticks) / float64(timescale); d > longest {
				longest = d
			}
		}
		snap.Segments = append(snap.Segments, seg)
	}
	for i := 0; i < sealed; i++ {
		add(first+i, partsPerSegment, true)
	}
	if openParts > 0 {
		add(first+sealed, openParts, false)
	}
	if seq > 0 {
		snap.HaveParts = true
		snap.NextPartSeq = seq + 1
	}
	snap.TargetSecs = int(longest + 0.999999)
	return snap
}

func meta() Meta {
	return Meta{
		SessionID:       "5a1b2c3d-4e5f-4a7b-8c9d-0e1f2a3b4c5d",
		ChannelID:       "chan_abc123",
		PartTargetMs:    500,
		SegmentTargetMs: 4000,
	}
}

func TestBuild(t *testing.T) {
	tests := []struct {
		name   string
		video  ring.Snapshot
		audio  *ring.Snapshot
		wantOK bool
		check  func(t *testing.T, got State)
	}{
		{
			name:   "complete segments, an open segment with two parts, and a preload hint",
			video:  videoSnapshot(41, 2, 4, 2, videoTimescale, partTicks),
			wantOK: true,
			check: func(t *testing.T, got State) {
				if got.MediaSequence != 41 {
					t.Fatalf("mediaSequence = %d, want the oldest listed segment (41)", got.MediaSequence)
				}
				if len(got.Video.Segments) != 3 {
					t.Fatalf("segments = %d, want 3", len(got.Video.Segments))
				}
				for i, want := range []bool{true, true, false} {
					if got.Video.Segments[i].Complete != want {
						t.Fatalf("segment %d complete = %v, want %v", i, got.Video.Segments[i].Complete, want)
					}
				}
				last := got.Video.Segments[2]
				if last.URI != nil || last.DurationSecs != nil {
					t.Fatalf("the open segment must carry neither uri nor duration: %+v", last)
				}
				if len(last.Parts) != 2 {
					t.Fatalf("open segment parts = %d, want 2", len(last.Parts))
				}
				// Sealed: 8 parts consumed, so the open segment holds
				// parts 9 and 10 and the hint names 11.
				if got.Video.Segments[2].Parts[0].URI != "part-9.m4s" {
					t.Fatalf("first open part uri = %q, want part-9.m4s", got.Video.Segments[2].Parts[0].URI)
				}
				if got.Video.PreloadHint == nil {
					t.Fatal("preloadHint missing")
				}
				if *got.Video.PreloadHint != (PreloadHint{MSN: 43, Part: 2, URI: "part-11.m4s"}) {
					t.Fatalf("preloadHint = %+v, want {43 2 part-11.m4s}", *got.Video.PreloadHint)
				}
				if got.Video.InitURI != "init.mp4" {
					t.Fatalf("initUri = %q", got.Video.InitURI)
				}
				if *got.Video.Segments[0].URI != "seg-41.m4s" {
					t.Fatalf("segment uri = %q, want seg-41.m4s", *got.Video.Segments[0].URI)
				}
				if got.Audio != nil {
					t.Fatalf("audio should be absent when no audio ring was passed: %+v", got.Audio)
				}
			},
		},
		{
			name:   "empty session before the first part",
			video:  ring.Snapshot{Timescale: videoTimescale, HasInit: true},
			wantOK: false,
		},
		{
			name:   "init segment written but no part yet",
			video:  ring.Snapshot{Timescale: videoTimescale, HasInit: true, Segments: nil},
			wantOK: false,
		},
		{
			name: "parts but no init segment",
			video: func() ring.Snapshot {
				s := videoSnapshot(0, 1, 2, 1, videoTimescale, partTicks)
				s.HasInit = false
				return s
			}(),
			wantOK: false,
		},
		{
			name:   "a sealed last segment hints part 0 of the next one",
			video:  videoSnapshot(7, 2, 4, 0, videoTimescale, partTicks),
			wantOK: true,
			check: func(t *testing.T, got State) {
				if *got.Video.PreloadHint != (PreloadHint{MSN: 9, Part: 0, URI: "part-9.m4s"}) {
					t.Fatalf("preloadHint = %+v, want {9 0 part-9.m4s}", *got.Video.PreloadHint)
				}
			},
		},
		{
			name:  "the audio twin is prefixed, every part independent, and its own msn space",
			video: videoSnapshot(41, 2, 4, 2, videoTimescale, partTicks),
			audio: func() *ring.Snapshot {
				s := videoSnapshot(12, 1, 2, 1, audioTimescale, audioTimescale/2)
				return &s
			}(),
			wantOK: true,
			check: func(t *testing.T, got State) {
				if got.Audio == nil {
					t.Fatal("audio track missing")
				}
				if got.Audio.InitURI != "audio-init.mp4" {
					t.Fatalf("audio initUri = %q", got.Audio.InitURI)
				}
				if *got.Audio.Segments[0].URI != "audio-seg-12.m4s" {
					t.Fatalf("audio segment uri = %q", *got.Audio.Segments[0].URI)
				}
				if got.Audio.Segments[0].Parts[1].URI != "audio-part-2.m4s" {
					t.Fatalf("audio part uri = %q", got.Audio.Segments[0].Parts[1].URI)
				}
				for _, seg := range got.Audio.Segments {
					for _, p := range seg.Parts {
						if !p.Independent {
							t.Fatalf("every audio part must be independent: %+v", p)
						}
					}
				}
				// Only the SECOND part of a video segment proves the
				// override is not blanket-true for both tracks.
				if got.Video.Segments[0].Parts[1].Independent {
					t.Fatal("a mid-segment video part must not claim independence")
				}
				if got.MediaSequence != 41 {
					t.Fatalf("mediaSequence = %d, want the VIDEO track's oldest (41)", got.MediaSequence)
				}
				if got.Audio.PreloadHint.URI != "audio-part-4.m4s" {
					t.Fatalf("audio preload hint = %+v", *got.Audio.PreloadHint)
				}
			},
		},
		{
			name:   "an audio ring with nothing in it yet leaves a video-only document",
			video:  videoSnapshot(41, 1, 4, 1, videoTimescale, partTicks),
			audio:  &ring.Snapshot{Timescale: audioTimescale, HasInit: true},
			wantOK: true,
			check: func(t *testing.T, got State) {
				if got.Audio != nil {
					t.Fatalf("audio = %+v, want absent", got.Audio)
				}
			},
		},
		{
			name: "a gap in the segment indices is refused, not rendered",
			video: func() ring.Snapshot {
				s := videoSnapshot(41, 2, 4, 1, videoTimescale, partTicks)
				s.Segments[1].Index = 44
				s.Segments[2].Index = 45
				return s
			}(),
			wantOK: false,
		},
		{
			name: "an incomplete segment before the live edge is refused",
			video: func() ring.Snapshot {
				s := videoSnapshot(41, 2, 4, 1, videoTimescale, partTicks)
				s.Segments[0].Sealed = false
				return s
			}(),
			wantOK: false,
		},
		{
			name: "targetDuration never lands below a listed segment's own duration",
			video: func() ring.Snapshot {
				s := videoSnapshot(0, 1, 8, 1, videoTimescale, partTicks) // 4.0s sealed
				s.TargetSecs = 1                                          // a stale/too-small running max
				return s
			}(),
			wantOK: true,
			check: func(t *testing.T, got State) {
				if got.TargetDurationSecs != 4 {
					t.Fatalf("targetDurationSecs = %v, want 4 (ceil of the 4.0s segment listed)", got.TargetDurationSecs)
				}
			},
		},
		{
			name: "a zero-tick part is floored rather than blanking the document",
			video: func() ring.Snapshot {
				s := videoSnapshot(0, 1, 2, 1, videoTimescale, partTicks)
				s.Segments[1].Parts[0].DurationTicks = 0
				return s
			}(),
			wantOK: true,
			check: func(t *testing.T, got State) {
				if got.Video.Segments[1].Parts[0].DurationSecs != minDurationSecs {
					t.Fatalf("durationSecs = %v, want the %v floor", got.Video.Segments[1].Parts[0].DurationSecs, minDurationSecs)
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			var audio *ring.Snapshot
			if tt.audio != nil {
				a := *tt.audio
				audio = &a
			}
			got, ok := Build(meta(), tt.video, audio)
			if ok != tt.wantOK {
				t.Fatalf("Build ok = %v, want %v", ok, tt.wantOK)
			}
			if !ok {
				return
			}
			if got.SessionID != meta().SessionID || got.ChannelID != meta().ChannelID {
				t.Fatalf("identity not carried through: %+v", got)
			}
			if tt.check != nil {
				tt.check(t, got)
			}
		})
	}
}

// TestGolden pins the exact document this package renders -- the same
// bytes tools/hls-edge/test/ll-state-remux-golden.test.mjs feeds to
// `parseLlState` and its playlist renderer. If this test and that one
// disagree, one side has drifted from the contract in
// tools/hls-edge/src/ll-state.js; run `go test ./internal/llstate
// -update-golden` only when the change is deliberate, and expect the
// Worker's cross-check to be the thing that says whether it is legal.
func TestGolden(t *testing.T) {
	video := videoSnapshot(41, 2, 4, 2, videoTimescale, partTicks)
	audioSnap := videoSnapshot(41, 2, 2, 1, audioTimescale, audioTimescale/2)
	state, ok := Build(meta(), video, &audioSnap)
	if !ok {
		t.Fatal("Build refused the golden snapshot")
	}

	// The bytes the HTTP surface actually writes (compact, one line),
	// re-indented for review. Comparing the indented form means the
	// golden file stays readable while still being the exact document
	// served, byte for byte, once whitespace is normalized.
	compact, err := json.Marshal(state)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var pretty bytes.Buffer
	if err := json.Indent(&pretty, compact, "", "  "); err != nil {
		t.Fatalf("indent: %v", err)
	}
	pretty.WriteString("\n")

	path := filepath.Join("testdata", "state-golden.json")
	if *updateGolden {
		if err := os.WriteFile(path, pretty.Bytes(), 0o644); err != nil {
			t.Fatalf("writing golden: %v", err)
		}
		t.Logf("wrote %s", path)
		return
	}
	want, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("reading golden (run with -update-golden to create it): %v", err)
	}
	if !bytes.Equal(pretty.Bytes(), want) {
		t.Fatalf("state.json drifted from the golden.\n--- got ---\n%s\n--- want ---\n%s", pretty.String(), want)
	}
}

// TestAudioTwinIsRenderedFromRealFragmenterOutput closes the gap the
// 2026-09-15 incident fell through. Every other test in this file (the
// golden included) hands Build a SYNTHETIC snapshot with 500ms audio
// parts, which is what the contract says and what the Worker's own
// cross-check then reads back -- so both sides agreed with each other
// while the actual producer, pipeline.AudioFragmenter, was emitting one
// part per 21ms AAC frame. Nothing in either repository compared the
// document to what the fragmenter really makes. This does: a real
// AudioFragmenter through a real ring into Build.
func TestAudioTwinIsRenderedFromRealFragmenterOutput(t *testing.T) {
	const partMs = 500
	const segmentMs = 4000
	const frameTicks = 1024 // aacenc.SamplesPerFrame

	audioRing := ring.New(6, audioTimescale)
	audioRing.SetInit([]byte("audio-init"))
	audioRing.SetClock(func() time.Time { return epoch })

	f := pipeline.NewAudioFragmenter(pipeline.AudioConfig{
		Timescale:       audioTimescale,
		PartDuration:    partMs * audioTimescale / 1000,
		SegmentDuration: segmentMs * audioTimescale / 1000,
	})

	// Twelve seconds of audio: three whole segments and a bit.
	var pts int64
	for i := 0; i < 12*audioTimescale/frameTicks; i++ {
		if frag := f.Push(pts, frameTicks, []byte{byte(i)}); frag != nil {
			audioRing.Push(frag)
		}
		pts += frameTicks
	}

	video := videoSnapshot(0, 2, 8, 2, videoTimescale, partTicks)
	audioSnap := audioRing.Snapshot()
	state, ok := Build(meta(), video, &audioSnap)
	if !ok {
		t.Fatal("Build refused a real audio ring")
	}
	if state.Audio == nil {
		t.Fatal("no audio twin")
	}

	// Parts are PART_MS, within one AAC frame. 0.021 -- one part per
	// frame -- is the bug.
	const frameSecs = float64(frameTicks) / audioTimescale
	target := float64(partMs) / 1000
	var parts int
	for _, seg := range state.Audio.Segments {
		for i, p := range seg.Parts {
			parts++
			last := i == len(seg.Parts)-1
			if p.DurationSecs > target+frameSecs {
				t.Fatalf("audio part %s is %.3fs, past the %.3fs target", p.URI, p.DurationSecs, target)
			}
			// Only the part a segment boundary cut short may be under
			// the target.
			if !last && p.DurationSecs < target {
				t.Fatalf("audio part %s is %.3fs, under the %.3fs target mid-segment", p.URI, p.DurationSecs, target)
			}
		}
	}
	if parts == 0 {
		t.Fatal("no audio parts at all")
	}

	// Segments hold the same order of magnitude of parts as the video
	// rendition's: 8 against 8, not 69 against 6.
	for _, seg := range state.Audio.Segments {
		if !seg.Complete {
			continue
		}
		if n := len(seg.Parts); n < 7 || n > 9 {
			t.Fatalf("audio segment %d holds %d parts at a %dms part target and a %dms segment target, want about %d",
				seg.MSN, n, partMs, segmentMs, segmentMs/partMs)
		}
	}

	// The preload hint names the part the fragmenter has not emitted
	// yet -- the very next sequence number, not one 20,000 ahead.
	hint := state.Audio.PreloadHint
	if hint == nil {
		t.Fatal("no audio preload hint")
	}
	lastSeg := state.Audio.Segments[len(state.Audio.Segments)-1]
	lastPart := lastSeg.Parts[len(lastSeg.Parts)-1]
	var lastSeq, hintSeq int
	if _, err := fmt.Sscanf(lastPart.URI, "audio-part-%d.m4s", &lastSeq); err != nil {
		t.Fatalf("parsing %q: %v", lastPart.URI, err)
	}
	if _, err := fmt.Sscanf(hint.URI, "audio-part-%d.m4s", &hintSeq); err != nil {
		t.Fatalf("parsing %q: %v", hint.URI, err)
	}
	if hintSeq != lastSeq+1 {
		t.Fatalf("preload hint %q follows the last published part %q by %d, want 1", hint.URI, lastPart.URI, hintSeq-lastSeq)
	}
}

// PART-TARGET is a promise about the MAXIMUM part duration, so a quiet
// source publishing parts longer than PART_MS must be reported honestly:
// the edge Worker times its blocking playlist reloads at three of these,
// and a stale 500ms times a viewer out on a healthy stream.
func TestBuild_PartTargetCoversTheLongestPart(t *testing.T) {
	// A static Chrome tab at ~1.4 frames/s: parts close on real frames,
	// about a second apart.
	oneSecond := uint32(videoTimescale)
	got, ok := Build(meta(), videoSnapshot(3, 1, 4, 2, videoTimescale, oneSecond), nil)
	if !ok {
		t.Fatal("Build refused a snapshot of second-long parts")
	}
	if got.PartTargetMs != 1000 {
		t.Fatalf("partTargetMs = %d against second-long parts, want 1000", got.PartTargetMs)
	}

	// An ordinary source never moves it off the configured value.
	got, ok = Build(meta(), videoSnapshot(3, 1, 4, 2, videoTimescale, partTicks), nil)
	if !ok {
		t.Fatal("Build refused an ordinary snapshot")
	}
	if got.PartTargetMs != 500 {
		t.Fatalf("partTargetMs = %d on a 500ms source, want the configured 500", got.PartTargetMs)
	}
}

// Every part duration here descends from the PUBLISHER's own access-unit
// timestamps, and the browser on the far side of the SFU is not a trusted
// input: two frames stamped an hour apart would otherwise become the
// edge's blocking-reload deadline for every viewer of the session. A part
// is never usefully longer than a segment, so the segment target is the
// ceiling.
func TestBuild_PartTargetIsBoundedByTheSegmentTarget(t *testing.T) {
	hostile := uint32(videoTimescale) * 3600 // an hour in one part
	got, ok := Build(meta(), videoSnapshot(3, 1, 1, 1, videoTimescale, hostile), nil)
	if !ok {
		t.Fatal("Build refused the snapshot")
	}
	if got.PartTargetMs != 4000 {
		t.Fatalf("partTargetMs = %d from an hour-long part, want it capped at the 4000ms segment target", got.PartTargetMs)
	}
	if got.PartTargetMs <= 0 {
		t.Fatal("parseLlState refuses a non-positive partTargetMs, which blanks the stream")
	}
}

// TestBuild_AudioSegmentInitURINeverAdvertisesBareInit is the Farol finding
// on PR #656: an audio ring whose segments still carry the ring-internal
// DefaultInitURI ("init.mp4" — what SetInit stores, and what a live audio
// ring stamped before SetNamedInit("audio-init.mp4")) must NEVER emit that
// bare name on state.json segments. The Worker prefers
// track.segments[0].initUri for the audio MAP, and ll-audio/init.mp4 is
// refused by nameBelongsToRung (audio rung requires the audio- prefix).
func TestBuild_AudioSegmentInitURINeverAdvertisesBareInit(t *testing.T) {
	video := videoSnapshot(41, 1, 2, 1, videoTimescale, partTicks)
	audio := videoSnapshot(41, 2, 2, 1, audioTimescale, audioTimescale/2)
	audio.InitURI = ring.DefaultInitURI // "init.mp4" — what a live audio ring stores
	for i := range audio.Segments {
		audio.Segments[i].InitURI = ring.DefaultInitURI
	}

	got, ok := Build(meta(), video, &audio)
	if !ok {
		t.Fatal("Build refused a snapshot whose audio segments carry init.mp4")
	}
	if got.Audio == nil {
		t.Fatal("audio track missing")
	}
	if got.Audio.InitURI != "audio-init.mp4" {
		t.Fatalf("audio track initUri = %q, want audio-init.mp4", got.Audio.InitURI)
	}
	for _, seg := range got.Audio.Segments {
		if seg.InitURI != "audio-init.mp4" {
			t.Fatalf("audio segment msn=%d initUri = %q, want audio-init.mp4 (never the ring's bare init.mp4)", seg.MSN, seg.InitURI)
		}
	}
}
